frontend/catalyst/compilation_pipelines.py

```diff
@@ -659,13 +688,21 @@ def _ensure_real_arguments_and_formal_parameters_are_compatible(self, function,
             next_action = CompiledFunction.typecheck(abstracted_axes, self.c_sig, r_sig)
 
         if next_action == TypeCompatibility.NEEDS_PROMOTION:
-            args = CompiledFunction.promote_arguments(self.c_sig, *args)
+            args = CompiledFunction.promote_arguments(self.c_sig, *dynamic_args)
         elif next_action == TypeCompatibility.NEEDS_COMPILATION:
             if self.user_typed:
                 msg = "Provided arguments did not match declared signature, recompiling..."
                 warnings.warn(msg, UserWarning)
             if not self.compiling_from_textual_ir:
-                self.mlir_module = self.get_mlir(*r_sig)
+                sig = r_sig
+                # Combine dynamic_args (in args) and r_sig (and keep the original order).
+                if static_argnums:
+                    sig = list(args)
+                    dynamic_indices = [idx for idx in range(len(args))\
+                        if idx not in static_argnums]
+                    for i, idx in enumerate(dynamic_indices):
+                        sig[idx] = r_sig[i]
```

Comment:
This does appear to duplicate the code in a different function, maybe we can factor it out?

Reply:
Yes, I will factor it out.